tokio/src/io/util/copy_bidirectional.rs

```diff
@@ -0,0 +1,116 @@
+        a: &'a mut A,
+        b: &'a mut B,
+        a_to_b: TransferState,
+        b_to_a: TransferState
+    }
+
+    fn transfer_one_direction<A, B>(
+        cx: &mut Context<'_>,
+        state: &mut TransferState,
+        r: &mut A,
+        w: &mut B
+    ) -> Poll<io::Result<u64>>
+        where A:  AsyncRead + AsyncWrite + Unpin + ?Sized, B: AsyncRead + AsyncWrite + Unpin + ?Sized
+    {
+        let mut r = Pin::new(r);
+        let mut w = Pin::new(w);
+
+        loop {
+            match state {
+                TransferState::Running(buf) => {
+                    let count = ready!(buf.poll_copy(cx, r.as_mut(), w.as_mut()))?;
+                    *state = TransferState::ShuttingDown(count);
+                }
+                TransferState::ShuttingDown(count) => {
+                    ready!(w.as_mut().poll_shutdown(cx))?;
+
+                    *state = TransferState::Done(*count);
+                }
+                TransferState::Done(count) => return Poll::Ready(Ok(*count)),
+            }
+        }
+    }
+
+    fn transpose<R, E>(r: Poll<Result<R,E>>) -> Result<Poll<R>, E> {
+        match r {
+            Poll::Pending => Ok(Poll::Pending),
+            Poll::Ready(Ok(r)) => Ok(Poll::Ready(r)),
+            Poll::Ready(Err(e)) => Err(e),
+        }
+    }
+
+    impl<'a, A, B> Future for CopyBidirectional<'a, A, B>
+        where A: AsyncRead + AsyncWrite + Unpin + ?Sized, B: AsyncRead + AsyncWrite + Unpin + ?Sized
+    {
+        type Output = io::Result<(u64, u64)>;
+
+        fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+            // Unpack self into mut refs to each field to avoid borrow check issues.
+            let CopyBidirectional { a, b, a_to_b, b_to_a } = &mut *self;
+
+            let a_to_b = transpose(transfer_one_direction(cx, a_to_b, &mut *a, &mut *b))?;
+            let b_to_a = transpose(transfer_one_direction(cx, b_to_a, &mut *b, &mut *a))?;
+
+            let a_to_b = ready!(a_to_b);
+            let b_to_a = ready!(b_to_a);
+
+            Ok((a_to_b, b_to_a)).into()
+        }
+    }
+
+    /// Copies data in both directions between `a` and `b`.
+    ///
+    /// This function returns a future which will read from both streams, and
+    /// write any data read to the opposing stream. The data transfer in one
+    /// direction (eg, `a` to `b`) will not be blocked if the opposite direction
+    /// blocks (e.g. if writing from `b` to `a` is blocked, `a` to `b` will not
+    /// be blocked).
+    ///
+    /// If an EOF is observed on one stream, [`shutdown()`] will be invoked on
+    /// the other, and reading from that stream will stop. Copying of data in
+    /// the other direction will continue.
+    ///
+    /// The future will complete successfully once both streams return EOF,
+    /// returning a tuple of the number of bytes copied from `a` to `b` and the
+    /// number of bytes copied from `b` to `a`, in that order.
```

Comment:
Add a note here that it returns once both streams return EOF _and the other side has been shut down_.

Reply:
I am also unsure about this

Reply:
How about this?

> The future will complete successfully once both directions of communication has been shut down. A direction is shut down when the reader reports EOF, at which point [`shutdown`][1] is called on the corresponding writer. When finished, it will return a tuple of the number of bytes copied from `a` to `b` and the number of bytes copied from `b` to `a`, in that order.

[1]: https://docs.rs/tokio/1/tokio/io/trait.AsyncWriteExt.html#method.shutdown